source/common/stats/thread_local_store.h

```diff
@@ -214,17 +217,19 @@ class ThreadLocalStoreImpl : Logger::Loggable<Logger::Id::stats>, public StoreRo
   const Stats::StatsOptions& statsOptions() const override { return stats_options_; }
 
 private:
+  template <class Stat> using StatMap = CharStarHashMap<Stat>;
+
   struct TlsCacheEntry {
-    std::unordered_map<std::string, CounterSharedPtr> counters_;
-    std::unordered_map<std::string, GaugeSharedPtr> gauges_;
-    std::unordered_map<std::string, TlsHistogramSharedPtr> histograms_;
-    std::unordered_map<std::string, ParentHistogramSharedPtr> parent_histograms_;
+    StatMap<CounterSharedPtr> counters_;
```

Comment:
And to your first question: right. We are using the stat itself as backing store for the name. This means in ThreadLocalStore we must avoid using operator[] on the maps with the temp string for lookup, as that string does not have a long lifetime. Instead we have to do a lookup with find() and then if it misses, do a separate insert, so there's an extra map lookup on a miss. OTOH there's less memory allocation and byte copying since we don't have to copy the string on the insert.